File: ruby/lib/svix/autoconfig_consumer.rb
# frozen_string_literal: true

require "uri"

require "svix/models/auto_config_sink_type"
require "svix/models/destination_in"
require "svix/models/destination_out"
require "svix/models/poller_v2_commit_in"
require "svix/models/sink_in_common"
require "svix/models/sink_status"
require "svix/models/subscribe_in"
require "svix/api_internal/endpoint_auto_config_deprecated"
require "svix/api_internal/destination_autoconfig"
require "svix/api_internal/message_pollerv2"

module Svix
  class AutoConfigConsumer
    def initialize(token, sink_in)
      content = AutoConfig.decode_token!(token)
      @app_id = content.fetch("app_id")
      @sink_id = content["endpoint_id"]
      @autoconfig_id = content["autoconfig_id"]
      @sink_in = sink_in
      @client = SvixHttpClient.new(
        content.fetch("token_plaintext"),
        URI(content.fetch("server_url"))
      )
    end

    def subscribe
      if @autoconfig_id
        destination = DestinationAutoconfig.new(@client).subscribe(
          @app_id,
          @autoconfig_id,
          sink_in_common_to_polling_destination(@sink_in)
        )
        @sink_id = destination.id
        destination
      else
        poller = AutoConfigSinkTypeConfig::Poller.deserialize(@sink_in.serialize)
        subscribe_in = SubscribeIn.new(
          "sink" => AutoConfigSinkType.new("config" => poller)
        )
        endpoint = EndpointAutoConfigDeprecated.new(@client).update(@app_id, @sink_id, subscribe_in)
        destination_out_from_v1_endpoint(endpoint)
      end
    end

    def receive(consumer_id, options = {})
      @sink_id ||= subscribe.id
      MessagePollerv2.new(@client).consumer_poll(
        @app_id,
        @sink_id,
        consumer_id,
        options
      )
    end

    def commit(consumer_id, offset, options = {})
      @sink_id ||= subscribe.id
      MessagePollerv2.new(@client).consumer_commit(
        @app_id,
        @sink_id,
        consumer_id,
        PollerV2CommitIn.new("offset" => offset),
        options
      )
    end

    private

    def destination_out_from_v1_endpoint(endpoint)
      DestinationOut.new(
        "id" => endpoint.id,
        "uid" => endpoint.uid,
        "status" => endpoint.disabled ? SinkStatus::DISABLED : SinkStatus::ENABLED,
        "current_iterator" => "",
        "created_at" => endpoint.created_at,
        "updated_at" => endpoint.updated_at,
        "batch_size" => 0,
        "max_wait_secs" => 0,
        "event_types" => endpoint.event_types,
        "channels" => endpoint.channels,
        "metadata" => endpoint.metadata,
        "config" => DestinationOutConfig::PollingEndpoint.new
      )
    end

    def sink_in_common_to_polling_destination(sink)
      attrs = {
        "config" => DestinationInConfig::PollingEndpoint.new
      }
      attrs["uid"] = sink.uid unless sink.uid.nil?
      attrs["event_types"] = sink.event_types unless sink.event_types.nil?
      attrs["channels"] = sink.channels unless sink.channels.nil?
      attrs["metadata"] = sink.metadata unless sink.metadata.nil?
      DestinationIn.new(attrs)
    end
  end
end

File: java/lib/src/main/java/com/svix/AutoConfigConsumer.java
package com.svix;

import com.svix.exceptions.ApiException;
import com.svix.internalapi.EndpointAutoConfigDeprecated;
import com.svix.internalapi.EndpointAutoconfig;
import com.svix.internalapi.DestinationAutoconfig;
import com.svix.internalapi.MessagePollerv2;
import com.svix.internalapi.MessagePollerv2ConsumerCommitOptions;
import com.svix.internalapi.MessagePollerv2ConsumerPollOptions;
import com.svix.models.AutoConfigSinkType;
import com.svix.models.AutoConfigSinkTypeConfig;
import com.svix.models.DestinationIn;
import com.svix.models.DestinationInConfig;
import com.svix.models.DestinationOut;
import com.svix.models.DestinationOutConfig;
import com.svix.models.EndpointOut;
import com.svix.models.PollerV2CommitIn;
import com.svix.models.PollerV2PollOut;
import com.svix.models.SinkInCommon;
import com.svix.models.DestinationStatus;
import com.svix.models.SubscribeIn;
import java.io.IOException;
import java.util.ArrayList;

public final class AutoConfigConsumer {
  private final String appId;
  private String sinkId;
  private final String autoconfigId;
  private final SinkInCommon sinkIn;
  private final Svix svix;

  public AutoConfigConsumer(final String token, final SinkInCommon sinkIn)
      throws AutoConfig.InvalidTokenException {
    AutoConfig.DecodedTokenContent content = AutoConfig.decodeToken(token);
    this.appId = content.getAppId();
    this.sinkId = content.getEndpointId();
    this.autoconfigId = content.getAutoconfigId();
    this.sinkIn = sinkIn;

    try {
      SvixOptions options = new SvixOptions();
      options.setServerUrl(content.getServerUrl());
      this.svix = new Svix(content.getTokenPlaintext(), options);
    } catch (IllegalArgumentException e) {
      throw new AutoConfig.InvalidTokenException(e);
    }
  }

  /** Registers this polling sink with Svix using the auto-config API. */
  public DestinationOut subscribe() throws IOException, ApiException {
    if (autoconfigId != null) {
      DestinationOut destination = new DestinationAutoconfig(svix.getHttpClient())
          .subscribe(appId, autoconfigId, sinkInCommonToPollingDestination(sinkIn));
      this.sinkId = destination.getId();
      return destination;
    }

    EndpointOut endpoint =
        new EndpointAutoConfigDeprecated(svix.getHttpClient()).update(appId, sinkId, new SubscribeIn()
            .sink(new AutoConfigSinkType(
                new AutoConfigSinkTypeConfig.Poller(sinkIn))));
    return destinationOutFromV1Endpoint(endpoint);
  }

  private String getSinkId() throws IOException, ApiException {
    if (sinkId != null) {
      // Already have the sink id from subscribe() or the v1 token
      return sinkId;
    }

    // Get the sink id from the autoconfig id (v2)
    String destId = new EndpointAutoconfig(svix.getHttpClient()).get(appId, autoconfigId).getDestId();
    if (destId == null || destId.isEmpty()) {
      throw new IllegalStateException(
          "autoconfig subscription is pending. Have you called subscribe()?");
    }
    return destId;
  }

  public PollerV2PollOut receive(final String consumerId,
      final MessagePollerv2ConsumerPollOptions options) throws IOException, ApiException {
    String resolvedSinkId = getSinkId();
    return new MessagePollerv2(svix.getHttpClient()).consumerPoll(appId, resolvedSinkId, consumerId,
        options);
  }

  public PollerV2PollOut receive(final String consumerId) throws IOException, ApiException {
    return receive(consumerId, new MessagePollerv2ConsumerPollOptions());
  }

  public void commit(final String consumerId, final long offset,
      final MessagePollerv2ConsumerCommitOptions options) throws IOException, ApiException {
    String resolvedSinkId = getSinkId();
    PollerV2CommitIn commitIn = new PollerV2CommitIn();
    commitIn.setOffset(offset);
    new MessagePollerv2(svix.getHttpClient()).consumerCommit(appId, resolvedSinkId, consumerId,
        commitIn, options);
  }

  public void commit(final String consumerId, final long offset) throws IOException, ApiException {
    commit(consumerId, offset, new MessagePollerv2ConsumerCommitOptions());
  }

  private static DestinationOut destinationOutFromV1Endpoint(final EndpointOut endpoint) {
    DestinationOut destination = new DestinationOut();
    destination.setId(endpoint.getId());
    destination.setUid(endpoint.getUid());
    destination.setStatus(
        Boolean.TRUE.equals(endpoint.getDisabled()) ? DestinationStatus.DISABLED : DestinationStatus.ENABLED);
    destination.setCurrentIterator("");
    destination.setCreatedAt(endpoint.getCreatedAt());
    destination.setUpdatedAt(endpoint.getUpdatedAt());
    destination.setBatchSize(0);
    destination.setMaxWaitSecs(0);
    if (endpoint.getEventTypes() != null) {
      destination.setEventTypes(new ArrayList<>(endpoint.getEventTypes()));
    }
    if (endpoint.getChannels() != null) {
      destination.setChannels(new ArrayList<>(endpoint.getChannels()));
    }
    destination.setMetadata(endpoint.getMetadata());
    destination.setConfig(new DestinationOutConfig.PollingEndpoint());
    return destination;
  }

  private static DestinationIn sinkInCommonToPollingDestination(final SinkInCommon sink) {
    DestinationIn destination = new DestinationIn()
        .config(new DestinationInConfig.PollingEndpoint())
        .uid(sink.getUid())
        .metadata(sink.getMetadata());
    if (sink.getEventTypes() != null) {
      destination.eventTypes(new ArrayList<>(sink.getEventTypes()));
    }
    if (sink.getChannels() != null) {
      destination.channels(new ArrayList<>(sink.getChannels()));
    }
    return destination;
  }
}
